### ISSUE TITLE
Optimize Krylov solver for batched calculations

### ISSUE THREAD

**Issue by vytautas-a** [pasqal-io/pyqtorch#295]
### Describe the feature

Currently (#294) batching in Krylov solver is implemented straightforwardly using a simple loop. This should be optimized to harness the power of broadcasted tensor operations.  

### Would you like to work on this issue?

None

**Comment by mlahariya** [comment]
@vytautas-a Are we starting to work on this? Do we also need to align with anton about the krylov solver already implemented?

**Comment by a-quelle** [comment]
> [@vytautas-a](https://github.com/vytautas-a) Are we starting to work on this? Do we also need to align with anton about the krylov solver already implemented?

We don't plan to do any batching, since that's mostly a machine learning thing. That said, as long as you're not storing gradients, doing the loop is more memory efficient, since you don't have to store all the kylov vectors for each batch index at the same time. If you start backpropagating, they will be stored in the computational graph anyway, and what you describe has no negative impact on memory consumption until you start implementing a custom backward pass (see the work in our repo for a custom implementation that causes a serious improvement).

**Comment by chMoussa** [comment]
> > [@vytautas-a](https://github.com/vytautas-a) Are we starting to work on this? Do we also need to align with anton about the krylov solver already implemented?
> 
> We don't plan to do any batching, since that's mostly a machine learning thing. That said, as long as you're not storing gradients, doing the loop is more memory efficient, since you don't have to store all the kylov vectors for each batch index at the same time. If you start backpropagating, they will be stored in the computational graph anyway, and what you describe has no negative impact on memory consumption until you start implementing a custom backward pass (see the work in our repo for a custom implementation that causes a serious improvement).

Hey @a-quelle sure we will check it out!

**Comment by chMoussa** [comment]
Hello @cburdine,

> Hi developers, I'm here for UnitaryHack and this looks like a good issue to tackle. Can you assign this issue to me?
Sure, you're assigned.

> Update: I have a few minor items in need of clarification:
> 
> * Are you looking for a solution that can balance memory usage versus batch parallelizability? (e.g., using [`torch.vmap()`](https://docs.pytorch.org/docs/stable/generated/torch.vmap.html) with a user-defined `chunk_size`? If not, I'll go with just batching along the last dimension.

Not necessarily a solution that balances, but seems what you proposed would be solving batching while going the extra mile to solve this issue. Feel free to try.

> * Above, you mention that batching should be done along the last dimension. Is this fully supported in the `integrate()` function of the `Krylov` subclass? If not, I'll go ahead and look at modifying the Krylov solver to accommodate trailing batch dimensions with respect to  `y`.

I think you will need to modify there too. Probably a few tweak to match dimensions.
